ludwig/utils/checkpoint_utils.py

```diff
@@ -65,28 +71,65 @@ def get_latest_checkpoint_path(directory: str) -> str:
 
 
 @DeveloperAPI
-class Checkpoint:
+class Checkpoint(ABC):
     """Save and restore model and optimizer states."""
 
     def __init__(
-        self, model: BaseModel, optimizer: Optional[Optimizer] = None, scheduler: Optional[LRScheduler] = None
+        self,
+        distributed: "DistributedStrategy",
+        model: "BaseModel",
+        optimizer: Optional[Optimizer] = None,
+        scheduler: Optional[LRScheduler] = None,
     ):
         """Constructor."""
+        self.distributed = distributed
         self.model = model
         self.optimizer = optimizer
         self.scheduler = scheduler
         self.global_step = 0
 
-    def restore(self, save_path: str, device: Optional[torch.device] = None) -> bool:
-        """Restore a state from a saved checkpoint.
+    def prepare(self, directory: str):
+        # create checkpoint directory if it doesn't
+        # already exist
+        mkdir(directory)
+
+    @abstractmethod
+    def load(self, save_path: str, device: Optional[torch.device] = None) -> bool:
+        pass
+
+    @abstractmethod
+    def get_state_for_inference(self, save_path: str, device: Optional[torch.device] = None) -> Mapping[str, Any]:
+        pass
+
+    @abstractmethod
+    def save(self, save_path: str, global_step: int):
+        pass
+
+    def _get_global_step(self, state: Dict[str, Any], save_path: str) -> int:
+        global_step = state.get("global_step")
+        if global_step is None:
+            # Legacy step detection for older checkpoint format which encoded the
+            # step number in the checkpoint filename.
+            return int(os.path.basename(save_path).split(".")[0])
+        return global_step
+
+
+@DeveloperAPI
+class CoordinatorCheckpoint(Checkpoint):
+    def prepare(self, directory: str):
+        if self.distributed.is_coordinator():
+            super().prepare(directory)
+
+    def load(self, save_path: str, device: Optional[torch.device] = None) -> bool:
+        """Load state from a saved checkpoint.
```

Comment:
With model parallelism, how does the model get loaded back? Is the idea that we load on the main coordinator device, then shard across devices again? If yes, is there a need to wrap the load call with `self.distributed.is_coordinator()`?

Reply:
Each worker loads is part of the model independently, so each worker needs to call it.